subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/DefaultToolchainSpec.java

```diff
@@ -78,14 +88,21 @@ public boolean equals(Object o) {
             return false;
         }
         DefaultToolchainSpec that = (DefaultToolchainSpec) o;
-        return Objects.equal(languageVersion.get(), that.languageVersion.get()) &&
-            Objects.equal(vendor.get(), that.vendor.get()) &&
-            Objects.equal(implementation.get(), that.implementation.get());
+        return Objects.equal(languageVersion.getOrNull(), that.languageVersion.getOrNull()) &&
+            Objects.equal(vendor.getOrNull(), that.vendor.getOrNull()) &&
+            Objects.equal(implementation.getOrNull(), that.implementation.getOrNull());
     }
 
     @Override
     public int hashCode() {
```

Comment:
Why does this type have `equals` and `hashCode`? Is is mutable, so it shouldn't be compared or end up in a collection.

Reply:
That's a good point, these should be removed.

Reply:
Unfortunately, it's not that straight-forward, as `JavaToolchainSpec` implementations, including `DefaultToolchainSpec` are used as keys in the toolchain cache:
https://github.com/gradle/gradle/blob/4ce52bc87e0bd8058688551f4737b212aa4e5bce/subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/JavaToolchainQueryService.java#L44

Reply:
Hmm... Are all the `JavaToolchainSpec` properties finalized on read? Then this may even make sense. Though it seems better to me if the keys would be some immutable data structures and we'd convert a spec to that immutable data structure before looking things up in the map.

Reply:
I agree that it is not ideal that we use mutable objects as keys, given that these objects come as arguments and not copied. At the same time, the current API does not seem to allow the users to get a hold of an actual `JavaToolchainSpec` instance. So corrupting the toolchain cache by mutating the keys externally seems to be very unlikely.

I agree that we probably should (in a separate PR) fix this by introducing something like this:
```java
public interface JavaToolchainSpecInternal extends JavaToolchainSpec {
    
    interface Key {}
    
    Key getKey();
    
}
```

Then use this key for actual caching of the toolchains. An alternative name that comes to mind is a `Snapshot`.

WDYT?

Reply:
We can leave this to a separate PR. The equals/hashCode is somewhat annoying since we also have `CurrentJvmToolchainSpec` and `SpecificInstallationToolchainSpec` which would then need some kind of special handling.

We can also not do this at all right now.

Reply:
Created an issue for this: https://github.com/gradle/gradle/issues/21598